crates/environ/src/fact/trampoline.rs

```diff
@@ -256,8 +402,282 @@ struct Memory<'a> {
+            }
+        }
+
+        self.finish()
+    }
+
+    fn compile_async_to_sync_adapter(
+        mut self,
+        adapter: &AdapterData,
+        start: FunctionId,
+        return_: FunctionId,
+        param_count: i32,
+        result_count: i32,
+    ) {
+        let enter = self.module.import_async_enter_call();
+        let exit = self
+            .module
+            .import_async_exit_call(None, adapter.lift.post_return);
+
+        self.flush_code();
+        self.module.funcs[self.result]
+            .body
+            .push(Body::RefFunc(start));
+        self.module.funcs[self.result]
+            .body
+            .push(Body::RefFunc(return_));
+        self.instruction(I32Const(
+            i32::try_from(adapter.lower.instance.as_u32()).unwrap(),
+        ));
+        self.instruction(I32Const(
+            i32::try_from(self.types[adapter.lift.ty].params.as_u32()).unwrap(),
+        ));
+        self.instruction(LocalGet(0));
+        self.instruction(LocalGet(1));
+        self.instruction(Call(enter.as_u32()));
+        self.module.exports.push((
+            adapter.callee.as_u32(),
+            format!("[adapter-callee]{}", adapter.name),
+        ));
+        self.instruction(I32Const(
+            i32::try_from(adapter.lower.instance.as_u32()).unwrap(),
+        ));
+        self.instruction(RefFunc(adapter.callee.as_u32()));
+        self.instruction(I32Const(
+            i32::try_from(adapter.lift.instance.as_u32()).unwrap(),
+        ));
+        self.instruction(I32Const(param_count));
+        self.instruction(I32Const(result_count));
+        self.instruction(I32Const(EXIT_FLAG_ASYNC_CALLER));
+        self.instruction(Call(exit.as_u32()));
+
+        self.finish()
+    }
+
+    fn compile_async_start_adapter(
+        mut self,
+        adapter: &AdapterData,
+        sig: &Signature,
+        param_globals: Option<&[u32]>,
+    ) {
+        let mut temps = Vec::new();
+        let param_locals = if let Some(globals) = param_globals {
+            for global in globals {
+                let ty = self.module.globals[usize::try_from(*global).unwrap()];
+
+                self.flush_code();
+                self.module.funcs[self.result]
+                    .body
+                    .push(Body::GlobalGet(*global));
+                temps.push(self.local_set_new_tmp(ty));
+            }
+            temps
+                .iter()
+                .map(|t| (t.idx, t.ty))
+                .chain(if sig.results_indirect {
+                    sig.params
+                        .iter()
+                        .enumerate()
+                        .map(|(i, ty)| (i as u32, *ty))
+                        .last()
+                } else {
+                    None
+                })
+                .collect::<Vec<_>>()
+        } else {
+            sig.params
+                .iter()
+                .enumerate()
+                .map(|(i, ty)| (i as u32, *ty))
+                .collect::<Vec<_>>()
```

Comment:
Ok I'll admit I'm still very lost as to what's going on here. My current understanding of async-start is that an async task calls it to get the parameters of the original async call. Here it looks like the adapter is performing the "translate from one component to another call", but I'm pretty confused why this signature's locals would ever be used. Why would a parameter be passed in if you're trying to learn the original parameters of the initial caller?

The globals bits make sense to me where the initial caller put some stuff in globals and that's read here, but I'm not sure how this works without the globals?

Reply:
The globals are gone, and I've added a bunch of comments; let me know if further clarification is needed.